routers/web/user/avatar.go

```diff
@@ -5,100 +5,49 @@
 package user
 
 import (
-	"errors"
-	"net/url"
-	"path"
-	"strconv"
 	"strings"
+	"time"
 
 	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/models/avatars"
 	"code.gitea.io/gitea/modules/context"
 	"code.gitea.io/gitea/modules/httpcache"
-	"code.gitea.io/gitea/modules/log"
-	"code.gitea.io/gitea/modules/setting"
 )
 
 func cacheableRedirect(ctx *context.Context, location string) {
-	ctx.Resp.Header().Set("Cache-Control", httpcache.GetCacheControl())
+	// here we should not use `setting.StaticCacheTime`, it is pretty long (default: 6 hours)
+	// we must make sure the redirection cache time is short enough, otherwise a user won't see the updated avatar in 6 hours
+	// it's OK to make the cache time short, it is only a redirection, and doesn't cost much to make a new request
+	httpcache.AddCacheControlToHeader(ctx.Resp.Header(), time.Minute)
```

Comment:
I don't agree to lowering this because it creates different caching behaviour of local vs. remote (Gravatar) avatars. Local avatars don't hit this code path so are still subject to `setting.StaticCacheTime`. I think we should either lower the default value or introduce a new config option for avatar cache time.

Reply:
It desn't creates different caching behaviour of local vs. remote (Gravatar) avatars, because we can not control or can not depend on Gravatar's cache behavior.

5 minutes is still too long. Why should we make a user can't see his new avatar in 5 minutes? And I do not think we should introduce so many config options, it only makes users more confused.

The `cacheableRedirect` was just introduced recently, there is no evidence any user benefits from it. The original issue was not likely related to it, either.

Reply:
But to be fair: How often do you change your avatar?
Once or twice a year? If at all?
So I can definitely understand caching it for a longer time.
If needed, the user can still force his cache to be cleared.

Reply:
But the 302 request doesn't cost much, it's not necessary to pay much attention to a fast 302 request. And even more, with this refactor, most local avatar images are responded by a `final` URL, they do not need the 302 redirection anymore.

The 302 redirection only occurs in some rare cases.

Reply:
Yes, we should get rid of this redirect, at least for the cases where the URL can be synchonously obtained, which is certainly possible for Gravatar (but not for some other federated services, IIRC).

So Gravatar should just render `img src="<gravatar-url>"`. Is this the case after this refactor?

Reply:
Yes, in most cases user's Gravatar image will be rendered as `img src="<gravatar-url>"`

The only cases the 302 redirections occurs:

1. In some cases we only have user name, then we use `/user/avatar/${name}/${size}` (currently no such case, there is no caller to `GenerateUserAvatarFastLink`, this function is for potential further cases.
2. In some cases we only have E-mail hash (eg: git commit log) AND `LibravatarService` is used, then we use `/avatar/${hash}`

I can not find more cases that need the 302 redirection. So the cases are rare enough.